src/data_structures/cgranges.rs

```diff
@@ -0,0 +1,293 @@
+//! Interval tree, a data structure for efficiently storing and searching intervals.
+//!
+//! This implementation is based on the sorted array version as described/given in
+//! https://github.com/lh3/cgranges / https://github.com/lh3/cgranges/blob/master/cpp/IITree.h
+//!
+//! It uses the same conventions as `crate::data_structures::interval_tree::IntervalTree`.
+//! Note that if you do not use the `IITree::from_iter` constructor, you have to call `index(&mut self)`
+//! first before `find()`-ing overlaps.
+//!
+//! # Example
+//! ```
+//! use bio::data_structures::cgranges::{IITree, Entry};
+//! use bio::utils::Interval;
+//!
+//! let mut tree = IITree::new();
+//! tree.insert(12..34, 0);
+//! tree.insert(0..23, 1);
+//! tree.insert(34..56, 2);
+//! // since we did at least one manual insert, we have to index the tree
+//! tree.index();
+//! let i1 = tree.find(22..25)[0];
+//! assert_eq!(i1.interval().start(), 0);
+//! assert_eq!(i1.interval().end(), 23);
+//! assert_eq!(i1.data(), 1);
+//!
+//! let tree = IITree::from_iter(vec![(12..34, 0), (0..23, 1), (34..56, 2)]);
+//! // no call to `index` needed here, since that happens in `from_iter` already
+//! let i2 = tree.find(22..25)[1];
+//! assert_eq!(i1.interval().start(), 12);
+//! assert_eq!(i1.interval().end(), 34);
+//! assert_eq!(i1.data(), 0);
+//!
+//! ```
+//!
+
+use std::cmp::min;
+
+use itertools::Itertools;
+
+use crate::utils::Interval;
+
+/// A `find` query on the interval tree does not directly return references to the intervals in the
+/// tree but wraps the fields `interval` and `data` in an `Entry`.
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct Entry<N: Ord + Clone + Copy, D> {
+    data: D,
+    interval: Interval<N>,
+    max: N,
+}
+
+impl<N: Ord + Clone + Copy, D> Entry<N, D> {
+    /// Get a reference to the data for this entry
+    pub fn data(&self) -> &D {
+        &self.data
+    }
+
+    /// Get a reference to the interval for this entry
+    pub fn interval(&self) -> &Interval<N> {
+        &self.interval
+    }
+}
+
+impl<N: Ord + Clone + Copy, D> Default for IITree<N, D> {
+    fn default() -> Self {
+        IITree {
+            entries: vec![],
+            max_level: 0,
+            indexed: false,
+        }
+    }
+}
+
+pub struct IITree<N: Ord + Clone + Copy, D> {
```

Comment:
Can you explain the name?

Reply:
... copied from the reference implementation. Open for Suggestions.

Reply:
`data_structures::interval_tree::ArrayBackedIntervalTree`

Reply:
Okay! I renamed ìnterval_tree.rs` to `avl_interval_tree.rs` and moved the cgranges code to `array_backed_interval_tree.rs`, moved those into the new `interval_tree` folder/module and adjusted public exports accordingly.

Reply:
I also changed the `ArrayBackedIntervalTree` code to return `Entry` with references to interval and data stored in the `ArrayBackedIntervalTree` instance instead of references to the `InternalEntry`, mimicking the `IntervalTree` interface. It does add some unneeded overhead I guess, though.